subsys/net/ip/Kconfig

```diff
@@ -159,11 +159,12 @@ config NET_SHELL_DYN_CMD_COMPLETION
 
 config NET_TC_TX_COUNT
 	int "How many Tx traffic classes to have for each network device"
-	default 1 if USERSPACE || USB_DEVICE_NETWORK
+	default 1 if USERSPACE || USB_DEVICE_NETWORK || NET_L2_IEEE802154
```

Comment:
IMO this should be restricted to cases where drivers are active that actually support timed/deferred TX. This is because having an extra thread is no free lunch and now we're forcing all drivers (even those that don't support deferred TX) into the extra overhead.

I propose that we add an intermediate (hidden) Kconfig variable NET_L2_IEEE802154_DELAYED_TX (which in the case of the nRF5 driver would be a vendor agnostic alias for NRF_802154_CAPABILITY_DELAYED_TX which should be deprecated then).

Depending on that configuration being selected by some driver, the additional thread will be added - otherwise immediate sending remains supported.

This makes sense independently of the proposed change because it will allow us to disable delayed TX (and the overhead incurred) even for drivers that support it.

Reply:
But what about regular `ieee802154_send()` operation (putting delayed TX aside) - in busy environments, CSMA and retransmissions will effectively block the calling thread, which is not a desired behavior in case of system work queue.

Perhaps though we shouldn't indeed force anyone to use extra thread (i. e. keep the ranges as they are).

Reply:
I'm not sure if this would have the desired effect.
The reason why I make this change has nothing to do with timed Tx, but instead it deals with the fact that ieee802154 API is synchronous.

[Zephyr documentation](https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#system-workqueue) states the following:
> A new workqueue can be justified if it is not possible for its work items to co-exist with existing system workqueue work items without an unacceptable impact; for example, if the new work items perform blocking operations that would delay other system workqueue processing to an unacceptable degree.

Now the fundamental problem here is that the ieee802154_radio API is designed to be synchronous. In particular, the `.tx()` function will block until one of the following condition is met:

For immediate Tx without ack request:
- transmit fails (busy channel or some other error occurs) -- may take hundreds of microseconds
- frame is successfully transmitted -- may take many milliseconds
If radio supports `IEEE802154_HW_TX_RX_ACK` capability and ack is requested
- frame is successfully transmitted and ack frame is received -- may take many milliseconds
- frame is successfully transmitted but ack is not recieved -- may take many milliseconds
If radio supports `IEEE802154_HW_TXTIME` and timed Tx is performed
- frame is successfully transmitted and (optionally) ack frame is received -- may take an undetermined amount of time, but in practice its up to tens of milliseconds.

This is true not only for nRF5 platform, but for all other implementers of the ieee802154 API. 

Users of the API also expect synchronous behavior, for example [OpenThread](https://github.com/zephyrproject-rtos/zephyr/blob/12e484cd4133c55b3ec31196c9bfc3aa02ef3bf0/modules/openthread/platform/radio.c#L337).
I suppose Zephyr IP stack also expects the behavior, but the errors are propagated through multiple layers of abstraction and I can't link a specific piece of code.
I don't see any way for the radio API implementation to notify the upper layers of any errors asynchronously after we return from `.tx()`.

Here is and (incomplete) list of other ieee802154_radio API implementations that block:
- [rf2xx](https://github.com/zephyrproject-rtos/zephyr/blob/12e484cd4133c55b3ec31196c9bfc3aa02ef3bf0/drivers/ieee802154/ieee802154_rf2xx.c#L688)
- [mcr20a](https://github.com/zephyrproject-rtos/zephyr/blob/12e484cd4133c55b3ec31196c9bfc3aa02ef3bf0/drivers/ieee802154/ieee802154_mcr20a.c#L1154C16-L1154C16)
- [kw41z](https://github.com/zephyrproject-rtos/zephyr/blob/12e484cd4133c55b3ec31196c9bfc3aa02ef3bf0/drivers/ieee802154/ieee802154_kw41z.c#L698C21-L698C29)

Reply:
@e-rk Thank you very much for that detailed analysis and bringing up this topic in the first place. It's a good thing that we discuss this. I agree that .tx() being blocking is not the root problem we want to solve. I'm just not yet convinced of the specific approach being proposed, as IMO:
 * It throws out the child with the bath water (i.e. we're just dealing time + simplicity for space + added complexity and I'm not convinced that this will be the best compromise for most users - to the contrary, I believe there are good arguments to not change the default unless we can show that we solve a practical and not just a theoretical problem for most (!) users. Working queues do not forbid blocking work on principle - only if it is a problem.
 * If we have that problem we don't have it for IEEE 802.15.4 only. That's why the CSMA argument is not convincing IMO as it applies to any shared medium (i.e. for most or maybe all PHYs supported by Zephyr, not only IEEE 802.15.4).

But I've not made up my mind about this yet. I'm actively thinking about this, so I can propose an alternative solution. I just believe we need to think more about this.

Reply:
@e-rk @rlubos Just to make that explicit: You DID convince me that the solution I proposed isn't a solution either...

Reply:
Having a dedicated workqueue for networking code (for delays/timeouts/retransmissions) is also an option. That way we'd free the system workqueue from any potentially blocking networking operations. Adding a thin wrapper over `k_work_schedule/reschedule_for_queue()` as a replacement for currently used `k_work_schedule/reschedule()` doesn't seem like a very complex task.

@jukkar What's your opinion on this, would it make sense to go that way?

Reply:
@rlubos Would that then be a configurable option so that the user still has the choice to NOT trade time for space?

Reply:
The traffic class support is implemented using a thread(s) so if we have 1 of them (like what is proposed here) then the sent packet will be handled by it. Not sure why we would need a separate workqueue.
But this could be made optional, so if the user does not care or does not have problems with current way of working (synchronous sending), then it could be allowed.

Reply:
@fgrandel I have another problem that the change I proposed solves, but didn't bring it up, because it applies only to nRF5 platforms: https://github.com/zephyrproject-rtos/zephyr/pull/61242#issue-1840297016

The gist of the problem is that we use a synchronous RPC protocol (Spinel) to serialize towards the network core. Since ieee802154_radio API is synchronous as well, that should be a perfectly valid thing to do. The problem comes when you combine that with the fact that serialization backend (ICMsg) also utilizes system workqueue to process incoming messages. This unfortunately causes deadlocks with the synchronous RPC protocol if the RPC request were made from the system workqueue context as well.

I think I will make a change to delegate sending of frames to a separate thread on nRF5 platform only for the time being, so it doesn't affect other platforms.

Reply:
> But this could be made optional,

@jukkar The asynchronous approach already *is* optional. Thinking again, I agree that a separate work queue wouldn't add much value then over the thread + fifo approach (which is maybe even a tad more lightweight). @rlubos I think I was a little blind there. 

But then leaving an extra thread off and optional by default is just the current solution which suggests that we don't need a change at all - at least not generally for IEEE 802.15.4.

> The problem comes when you combine that with the fact that serialization backend (ICMsg)

@e-rk That convinces me even more that we're solving a very special problem with a much too general solution.

Why don't you simply require the separate TX thread for this very specific combination of configurations then rather than wanting to introduce it for the whole IEEE 802.15.4 protocol family as a default? You should still introduce a separate (hidden) intermediate Kconfig variable for improved separation of concerns which you select from a local Kconfig where the Spinel protocol is being configured.

Reply:
My assumption is that blocking calls should not be made from the system workqueue context. At least that's what the in-tree Zephyr code should strive for IMO. Therefore in my view delegating `.tx()` to separate thread solves what I consider to be the root cause and also happens to help me with other problems I encountered that lead me here.

I am still convinced that using ieee802154_radio API from system workqueue context is the wrong thing to do, as it adds considerable latency for most users of `k_work` API, keeping the CPU idle when it could do some useful work. However I understand that this change also has considerable impact on memory consumption, so I am fine with just handling the specific configuration case that I want to use for the time being.

Reply:
@rlubos That potential for deadlock via the work-q should still be kept in mind. It cannot be solved by making the time the queue is blocked shorter. This will just reduce the probability of deadlock but not the bug. (It's just less obvious and much harder to debug then as a general problem.) It's not the first time that TX/RX dependencies bite us, do you remember the deadlock lock situation I constructed some time ago. What @e-rk describes is very similar and completely unrelated to the main problem being discussed here.

Reply:
@e-rk 

> At least that's what the in-tree Zephyr code should strive for IMO.

It's good that you can see that we need to choose the lesser of two evils. Backwards compatibility and space trade offs are very strong arguments on embedded systems. That said: You problem needs to be solved so I'm happy that the proposed lightweight solution seems to solve it.

>  also happens to help me with other problems I encountered

Please make these problems explicit if you want us to consider them as well. We may be convinced after all. :-)